### PR TITLE
improve scheduling performance on batch jobs

### LINKED ISSUE #268: [Features] improve scheduling performance on batch jobs
**Is this a BUG REPORT or FEATURE REQUEST?**:
/kind feature
More like a feature improment
> /kind feature

**What happened**:
When the cluster idle resource can not satisfy some big job, the big job still holds the resource that allocated to it. This leads to this part of resource can not be used by other submitted job.

Is it better to release the resources of the tasks which are not bound to node as the idle resources of the cluster can not satisfy the whole job?
Currently, it often occurred that 2 jobs were pending even the idle resources of the cluster can satisfy the last submitted one.

**What you expected to happen**:
When the idle resource of cluster can not satisfy the job, the resource that already allocated to job should be released for other submitted job.

**How to reproduce it (as minimally and precisely as possible)**:
Suppose the idle resources of the cluster is: 100 cores, 100GB mem
1, submit 1st job with 4 pods, each req: 30 cores, 40GB mem
2, submit 2nd job with 4 pods, each req: 20 cores, 20GB mem
Both the 1st and 2nd job suspend even the idle resource of cluster can satisfy the 2nd job.

**Environment**:
- Volcano Version: master branch

### PR DESCRIPTION
**Pull request by william-wang**
**What this PR does / why we need it**:
This PR aims to make the big job release allocated resource for other submitted job when all idle resource of cluster can not satisfy the big job.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

Fixes https://github.com/volcano-sh/volcano/issues/268

### CODE REVIEW

File: pkg/scheduler/actions/allocate/allocate.go [hunk 1]
```diff
@@ -162,6 +164,12 @@ func (alloc *allocateAction) Execute(ssn *framework.Session) {
 				if err := ssn.Allocate(task, node.Name); err != nil {
 					glog.Errorf("Failed to bind Task %v on %v in Session %v, err: %v",
 						task.UID, node.Name, ssn.UID, err)
+				} else {
+					// store the allocate operation on task for rollback in the future
+					if err := stmt.Allocate(task, "allocate"); err != nil {
```
Comment (k82cn) [inline pkg/scheduler/actions/allocate/allocate.go]: let's put all related logic into `stmt.Allocate` and `stmt.Commit` when job is ready.

Comment (william-wang) [reply pkg/scheduler/actions/allocate/allocate.go]: okay, it make sense.

File: pkg/scheduler/framework/statement.go [hunk 2]
```diff
@@ -194,7 +196,117 @@ func (s *Statement) unpipeline(task *api.TaskInfo) error {
...
+	glog.V(3).Info("Allocating operations ...")
+	s.operations = append(s.operations, operation{
+		name: "allocate",
+		args: []interface{}{task, hostname},
+	})
+
+	return nil
+}
+
+func (s *Statement) allocate(task *api.TaskInfo, hostname string) error {
```
Comment (hex108) [inline pkg/scheduler/framework/statement.go:253]: Is it same as `ssn.dispatch()`? Could we reuse it?

File: pkg/scheduler/framework/statement.go [hunk 3]
```diff
@@ -194,7 +196,117 @@ func (s *Statement) unpipeline(task *api.TaskInfo) error {
 	return nil
 }
 
-// Discard operation for evict and pipeline
+// Save the allocate operation on task
+func (s *Statement) Allocate(task *api.TaskInfo, hostname string) error {
```
Comment (hex108) [inline pkg/scheduler/framework/statement.go:199]: Most of code is same as `ssn.Allocate()`, could we deduplicate them, e.g. move same code to a new function?

File: pkg/scheduler/framework/statement.go (hunk 2 above)
Comment (william-wang) [reply pkg/scheduler/framework/statement.go:253]: Great comments. The answer is currently the allocation func in statement is an enhancement compared to the allocation.go file.  Disscussed with Klaus, the principle is we will only use the Allocate  func in statement. The func in allocte.go will be deprecated in the future.

File: pkg/scheduler/framework/statement.go (hunk 3 above)
Comment (william-wang) [reply pkg/scheduler/framework/statement.go:199]: We will finally remove the Allocate func in allocate.go. For now there is still other places using it.

**Comment by hex108** [comment]
Last question:  would big jobs starve?

**Comment by william-wang** [comment]
> Last question: would big jobs starve?

Yes. The starvation will happen if small job is submitted continuosly.  The starvation issue is being resolved in another project planed by Klaus. This ticket is only focusing on the pending issue.

**Comment by hex108** [comment]
/lgtm

Could you please squash these commits to one commit?

**Comment by william-wang** [comment]
> /lgtm
> 
> Could you please squash these commits to one commit?

Sure. I just squash these commits. Thanks.

**Comment by k82cn** [comment]
>  The starvation issue is being resolved in another project planed by Klaus.

The starvation issue will be handled in https://github.com/volcano-sh/volcano/issues/13

**Comment by zionwu** [comment]
Will this PR cherry-pick to kube-batch repo?

**Pull request merged**